composer/trainer/trainer.py

```diff
@@ -1662,14 +1661,19 @@ def _train_batch(self, use_grad_scaling: bool):
                 if self.adaptive_gradient_accumulation and _is_cuda_oom(e):
                     log.debug((f"Rank {dist.get_global_rank()} OOM'd."))
                     should_handle_cuda_oom = 1
-                elif 'Timed out' in str(e):
-                    # Catch timeout errors and only reraise if we did not encounter OOM on other ranks. Error
-                    # is likely transient if one rank OOMed, it likely did not reach a barrier. Note that if we
-                    # catch non-transient timeout errors they will be later reraised if no rank OOMed.
-                    caught_timeout_error = e
                 else:
                     raise
 
+            # Use monitored barrier to error on deadlock. If one rank OOMs and another doesn't and gets stuck
+            # on a dist reduction in gradient syncronization, the monitored barrier will fail after the timeout.
+            try:
+                dist.monitored_barrier(timeout=datetime.timedelta(seconds=300))
```

Comment:
5 minutes seems....like a lot? Curious if we could shorten the time for a better UX for debugging? Would it make sense to set the timeout dynamically (say, `min(300, max(10, 0.5 * forward_backwards_pass_time))`)?